example(plate-balancing) : Add working plate balancing example with external c3 repo

Meow404 opened the pull request:
Fixes DAIRLab/c3#11
Dependency https://github.com/DAIRLab/c3/pull/10

- [ ] Waiting on https://github.com/DAIRLab/c3/pull/10 to be merged 
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/DAIRLab/dairlib/376)
<!-- Reviewable:end -->

Meow404 committed:
feat: inital commit -- bring plate balancing example from robust_c3

Meow404 committed:
fix: removed c3 dependency from franka_c3_controller.cc

Meow404 committed:
style: refactor

Meow404 committed:
dependencies: remove unnecessary dependencies

Meow404 committed:
doc: update readme

Meow404 committed:
fix: update .bazelrc

Meow404 committed:
fix: add c3 lcmtypes to lcm-spy

Meow404 committed:
fix: update to incorporate changes in c3

Meow404 committed:
fix: use update force visualization

Meow404 committed:
feat: use trajectory generator in C3

Meow404 committed:
fix: refactory and remove c3 state sender

Meow404 committed:
recfactor: cleanup

Meow404 committed:
fix: use static c3 library

ebianchi committed:
Include workspace and input limits as constraints in C3 QP step

juanmed commented:
Hi. Thanks for open sourcing your impressive work. I am working on this branch to reproduce the plate-balancing work, and was wondering if you accept fixes here? These are just a few missing dependencies https://github.com/juanmed/dairlib/commit/3d90389965b473459c0670e6f7cb86a9865851f9

Meow404 commented:
> Hi. Thanks for open sourcing your impressive work. I am working on this branch to reproduce the plate-balancing work, and was wondering if you accept fixes here? These are just a few missing dependencies [juanmed@3d90389](https://github.com/juanmed/dairlib/commit/3d90389965b473459c0670e6f7cb86a9865851f9)

Hi, Yes we accept fixes. Feel free to file bugs if you find any. 

I think you find the .yaml files for the plate balancing are included in this branch,. You should take a look at https://github.com/DAIRLab/c3 for our work related to C3 and https://github.com/DAIRLab/dairlib/tree/sampling_based_c3_PR2_copy I believe has the latest work for sampling based C3.

Meow404 committed:
feat: add lcmt systems

xuanhien070594 committed:
Resolve merge conflicts with main (#386)

* Resolve merge conflicts with main

* Remove obsolete codes

* Remove accidentally commited files

* Add external force tracking result to lcm

* Add back the deleted files in Cassie example

* Add back deleted files in pydairlib/cassie

* Revert code formatting

* Revert code formatting

* Revert code formatting

Meow404 committed:
fix: refactor after rebase on sampling_based_c3_public

Meow404 committed:
todo: add todo for robot_lcm_systems.h

Meow404 committed:
fix: pmd typo

Meow404 committed:
fix: minor fixes

Meow404 committed:
fix: revert changes in config

Meow404 committed:
feat: add delay to c3 controller

xuanhien070594 committed:
Integrate C3+ into push_anything branch (#390)

* Integrate new C3 algorithm

* Tune new C3 parameters

* Fix bug that wrongly computes the size of z and delta vectors.

* Remove unnecessary use of epsilon in new c3 projection step; update logging script to accept folder path as argument and modify log directory creation; add progress parameters YAML file for new c3

* Add log directory to .gitignore to prevent logging files from being tracked

* Minor changes

* Add documentation and address comments

* Refactor C3 solver structure:
- Introduce BaseC3 class, replace C3 with BaseC3 in related files.
- Enhances modularity and maintainability of all functions in BaseC3 class.
- Change name of C3NextGen class to C3Plus

* Update C3 solver implementation:
- Modify BaseC3 constructor to accept a function for calculating z size.
- Adjust C3Plus to disable parallelization and utilize the new z size calculation function.
- Clean up .gitignore by removing the logs directory entry.
- Improve documentation in C3Plus header for clarity on C3+ formulation.

* Clean up codes

* Replace all NEXTGEN name by C3+

* Minor changes

* Add C3+ parameters to jacktoy and push-anything examples

* Add missing placeholders in sampling_c3_options.yaml of push-anything example

* Change class name BaseC3 to C3Base

* Remove unnecessary small functions and remove obsolete SolveRobustSingleProjection functions

* Modify C3Base constructor to accept z_size and add overloaded constructor for default z_size.

* Use fully vectorized code in c3plus projection step

* Throw error if any weights of the cost matrix U are negative.

Meow404 committed:
feat: use external c3 repo for sampling example

xuanhien070594 committed:
Add reconstructed meshes of few objects, tune parameters of C3+ for push_t example

Meow404 committed:
feat: update commit for c3

Meow404 committed:
fixes after rebase

Meow404 committed:
feat: add config for C3+

Meow404 committed:
fix: use of correct N

juanmed commented:
@Meow404 Hi, I was trying out your work in this branch, looks awesome. As of the latest commit here, this is the behavior I get and was wondering if your get the same? 


https://github.com/user-attachments/assets/84cb23aa-c251-4a5d-9006-def938696d61


Asking because I expected the robot arm to 1) slide the tray out of the supports, 2) lift the tray and 3) slide the tray again onto the supports, but seems to do only 1) and 2).

Meow404 commented:
> @Meow404 Hi, I was trying out your work in this branch, looks awesome. As of the latest commit here, this is the behavior I get and was wondering if your get the same?
> 
>  slide-tray-2025-08-28_01.58.40.mp4 
> Asking because I expected the robot arm to 1) slide the tray out of the supports, 2) lift the tray and 3) slide the tray again onto the supports, but seems to do only 1) and 2).

Hi @juanmed, Nice to hear from you. We are discovering that the example does work a little differently based on your computer specs. I've tried the same example on three different computers with different results, but this can be overcome by tuning the parameters a little in this file `examples/plate-balancing/config/C3/controller-c3plus-options/2_supports.yaml`. You would want to start by trying to modify `w_[Q/R/G]` first, then maybe try `g_lambda`, `g_eta`. Let us know if you need any help.

juanmed commented:
@Meow404 Aha that makes sense, will play with the params there. Thanks for the advice!

(Btw I had to focus on some [background](https://github.com/RobotLocomotion/drake/pull/23197) work before retaking plate balancing again)

Meow404 committed:
fix: update c3 commit
